# Problem 1
Load Iris data. Use data transformations a classification algorithm of your choice to achieve greater than 98% accuracy in cross-validation (cv = 3).

In [24]:
import numpy as np
from sklearn import datasets
iris = datasets.load_iris()
X = iris.data 
y = iris.target
np.random.seed(42)

In [25]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import GridSearchCV

param_grid = [{'weights': ["uniform", "distance"], 'n_neighbors': [8,9,10,11,12,13,14,15]}]

knn_clf = KNeighborsClassifier()
grid_search = GridSearchCV(knn_clf, param_grid, cv=3, verbose=3, n_jobs=-1,scoring="accuracy")
grid_search.fit(X, y)


Fitting 3 folds for each of 16 candidates, totalling 48 fits
[CV] n_neighbors=8, weights=uniform ..................................
[CV] n_neighbors=8, weights=uniform ..................................
[CV] n_neighbors=8, weights=uniform ..................................
[CV] n_neighbors=8, weights=distance .................................
[CV] n_neighbors=8, weights=distance .................................
[CV] n_neighbors=8, weights=distance .................................
[CV] n_neighbors=9, weights=uniform ..................................
[CV]  n_neighbors=8, weights=uniform, score=0.9803921568627451, total=   0.0s
[CV]  n_neighbors=8, weights=uniform, score=0.9803921568627451, total=   0.0s
[CV] n_neighbors=9, weights=uniform ..................................
[CV] n_neighbors=9, weights=uniform ..................................
[CV]  n_neighbors=8, weights=uniform, score=0.9791666666666666, total=   0.0s
[CV] n_neighbors=9, weights=distance .............................

[Parallel(n_jobs=-1)]: Done  16 out of  48 | elapsed:    0.1s remaining:    0.1s
[Parallel(n_jobs=-1)]: Done  48 out of  48 | elapsed:    0.1s finished


GridSearchCV(cv=3, error_score='raise',
       estimator=KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=5, p=2,
           weights='uniform'),
       fit_params=None, iid=True, n_jobs=-1,
       param_grid=[{'weights': ['uniform', 'distance'], 'n_neighbors': [8, 9, 10, 11, 12, 13, 14, 15]}],
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring='accuracy', verbose=3)

In [26]:
from sklearn.model_selection import cross_val_score
np.mean(cross_val_score(grid_search, X, y, cv=3, scoring="accuracy"))

Fitting 3 folds for each of 16 candidates, totalling 48 fits
[CV] n_neighbors=8, weights=uniform ..................................
[CV] n_neighbors=8, weights=uniform ..................................
[CV] n_neighbors=8, weights=uniform ..................................
[CV] n_neighbors=8, weights=distance .................................
[CV] n_neighbors=8, weights=distance .................................
[CV] n_neighbors=8, weights=distance .................................
[CV] n_neighbors=9, weights=uniform ..................................
[CV]  n_neighbors=8, weights=uniform, score=0.9393939393939394, total=   0.0s
[CV]  n_neighbors=8, weights=uniform, score=0.9696969696969697, total=   0.0s
[CV]  n_neighbors=9, weights=uniform, score=0.9696969696969697, total=   0.0s
[CV] n_neighbors=9, weights=uniform ..................................
[CV] ........ n_neighbors=8, weights=uniform, score=1.0, total=   0.0s
[CV] n_neighbors=9, weights=uniform ..............................

[Parallel(n_jobs=-1)]: Done  16 out of  48 | elapsed:    0.1s remaining:    0.1s
[Parallel(n_jobs=-1)]: Done  48 out of  48 | elapsed:    0.1s finished


[CV] n_neighbors=8, weights=uniform ..................................
[CV] n_neighbors=8, weights=uniform ..................................
[CV] n_neighbors=8, weights=uniform ..................................
[CV] n_neighbors=8, weights=distance .................................
[CV] n_neighbors=8, weights=distance .................................
[CV] n_neighbors=8, weights=distance .................................
[CV] n_neighbors=9, weights=uniform ..................................
[CV]  n_neighbors=8, weights=uniform, score=0.9393939393939394, total=   0.0s
[CV] ........ n_neighbors=8, weights=uniform, score=1.0, total=   0.0s
[CV] ........ n_neighbors=8, weights=uniform, score=1.0, total=   0.0s
[CV] n_neighbors=9, weights=uniform ..................................
[CV] n_neighbors=9, weights=uniform ..................................
[CV] n_neighbors=9, weights=distance .................................
[CV]  n_neighbors=8, weights=distance, score=0.9393939393939394, total

[Parallel(n_jobs=-1)]: Done  16 out of  48 | elapsed:    0.0s remaining:    0.1s
[Parallel(n_jobs=-1)]: Done  48 out of  48 | elapsed:    0.1s finished


[CV] n_neighbors=8, weights=uniform ..................................
[CV] n_neighbors=8, weights=uniform ..................................
[CV] n_neighbors=8, weights=uniform ..................................
[CV] n_neighbors=8, weights=distance .................................
[CV] n_neighbors=8, weights=distance .................................
[CV] n_neighbors=8, weights=distance .................................
[CV] n_neighbors=9, weights=uniform ..................................
[CV]  n_neighbors=8, weights=uniform, score=0.9722222222222222, total=   0.0s
[CV] ........ n_neighbors=8, weights=uniform, score=1.0, total=   0.0s
[CV] n_neighbors=9, weights=uniform ..................................
[CV] n_neighbors=10, weights=uniform .................................
[CV]  n_neighbors=8, weights=uniform, score=0.8484848484848485, total=   0.0s
[CV] n_neighbors=9, weights=uniform ..................................
[CV] n_neighbors=9, weights=distance ..........................

[Parallel(n_jobs=-1)]: Done  16 out of  48 | elapsed:    0.0s remaining:    0.1s
[Parallel(n_jobs=-1)]: Done  48 out of  48 | elapsed:    0.1s finished


0.9869281045751634

**The accuracy in cross-validation (cv=3) is 98.6928%**

# Problem 2 
Load MNIST data set. Use data transformations a classification algorithm(s) of your choice to achieve greater than 95% accuracy in cross-validation (cv = 3). There are many ways to achieve, so there are many correct answers. Use only algorithms that we studied. KNN is forbidden! Hint: Data is big, look for more efficient algorithms. Start with simple models, make it complex if needed. On you computer the code should run less than 30 minutes. I have a fast computer, it should run less 30 minutes on my as well.

In [2]:
from sklearn.datasets import fetch_mldata
mnist = fetch_mldata('MNIST original')
X = mnist.data
y = mnist.target

In [3]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import cross_val_score


X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.5, random_state=42)

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)

/anaconda3/lib/python3.6/site-packages/sklearn/utils/validation.py:475: DataConversionWarning: Data with input dtype uint8 was converted to float64 by StandardScaler.
  warnings.warn(msg, DataConversionWarning)


In [7]:
from sklearn.model_selection import RandomizedSearchCV
from sklearn.ensemble import RandomForestClassifier
from scipy.stats import randint

param_distribs = {
        'n_estimators': randint(low=10, high=30),
        #'max_features': randint(low=15, high=20),
    }

forest_reg = RandomForestClassifier(random_state=42)
# Set the number of iterations as 10
rnd_search = RandomizedSearchCV(forest_reg, param_distributions=param_distribs,
                                n_iter=10, n_jobs=1, cv=5, scoring='accuracy', random_state=42)
rnd_search.fit(X_train_scaled, y_train)

RandomizedSearchCV(cv=5, error_score='raise',
          estimator=RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=1,
            oob_score=False, random_state=42, verbose=0, warm_start=False),
          fit_params=None, iid=True, n_iter=10, n_jobs=1,
          param_distributions={'n_estimators': <scipy.stats._distn_infrastructure.rv_frozen object at 0x1a1144dcc0>},
          pre_dispatch='2*n_jobs', random_state=42, refit=True,
          return_train_score='warn', scoring='accuracy', verbose=0)

In [8]:
import numpy as np
np.mean(cross_val_score(rnd_search, X_train_scaled, y_train, cv=3, scoring="accuracy"))

0.9515429169578525

# Problem 3 
Load CA housing data set. Use data transformations a regression algorithm(s) of your choice to achieve MSE less than $65,000 in predicting housing prices in cross-validation (cv = 3). There are many ways to achieve, so there are many correct answers. Use only algorithms that we studied. On you computer the code should run less than 30 minutes. I have a fast computer, it should run less 30 minutes on my as well.

In [27]:
from six.moves import urllib
import os
DOWNLOAD_ROOT = "https://raw.githubusercontent.com/ageron/handson-ml/master/"
HOUSING_PATH = os.path.join("datasets", "housing")
HOUSING_URL = DOWNLOAD_ROOT + "datasets/housing/housing.tgz"
import pandas as pd
housing = pd.read_csv('/Users/vannyzhang/Desktop/Machine Learning/Data/datasets/housing/housing.csv')
housing.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   
2    -122.24     37.85                52.0       1467.0           190.0   
3    -122.25     37.85                52.0       1274.0           235.0   
4    -122.25     37.85                52.0       1627.0           280.0   

   population  households  median_income  median_house_value ocean_proximity  
0       322.0       126.0         8.3252            452600.0        NEAR BAY  
1      2401.0      1138.0         8.3014            358500.0        NEAR BAY  
2       496.0       177.0         7.2574            352100.0        NEAR BAY  
3       558.0       219.0         5.6431            341300.0        NEAR BAY  
4       565.0       259.0         3.8462            342200.0        NEAR BAY

In [28]:
from sklearn.model_selection import train_test_split
from sklearn.model_selection import StratifiedShuffleSplit

train_set, test_set = train_test_split(housing, test_size=0.2, random_state=42)
housing["income_cat"] = np.ceil(housing["median_income"] / 1.5)
housing["income_cat"].where(housing["income_cat"] < 5, 5.0, inplace=True)
split = StratifiedShuffleSplit(n_splits=1, test_size=0.2, random_state=42)

for train_index, test_index in split.split(housing, housing["income_cat"]): 
    strat_train_set = housing.loc[train_index]
    strat_test_set = housing.loc[test_index]
    
housing = strat_train_set.copy()

In [29]:
housing.shape

(16512, 11)

In [30]:
housing = strat_train_set.drop("median_house_value", axis=1)
housing_labels = strat_train_set["median_house_value"].copy()

housing_num = housing.drop('ocean_proximity', axis=1)

from sklearn.pipeline import Pipeline
from sklearn.preprocessing import Imputer
from sklearn.preprocessing import StandardScaler

In [31]:
from sklearn.base import BaseEstimator, TransformerMixin

class DataFrameSelector(BaseEstimator, TransformerMixin):
    def __init__(self, attribute_names):
        self.attribute_names = attribute_names
    def fit(self, X, y=None):
        return self
    def transform(self, X):
        return X[self.attribute_names].values

In [32]:
from sklearn.base import TransformerMixin 
class MyLabelBinarizer(TransformerMixin):
    def __init__(self, *args, **kwargs):
        self.encoder = LabelBinarizer(*args, **kwargs)
    def fit(self, x, y=0):
        self.encoder.fit(x)
        return self
    def transform(self, x, y=0):
        return self.encoder.transform(x)

In [33]:
housing_cat = housing[['ocean_proximity']]

In [34]:
from future_encoders import OrdinalEncoder
ordinal_encoder = OrdinalEncoder()

housing_cat_encoded = ordinal_encoder.fit_transform(housing_cat)

In [35]:
from sklearn.preprocessing import LabelBinarizer
encoder = LabelBinarizer()
housing_cat_1hot = encoder.fit_transform(housing_cat)
housing_cat_1hot

array([[1, 0, 0, 0, 0],
       [1, 0, 0, 0, 0],
       [0, 0, 0, 0, 1],
       ...,
       [0, 1, 0, 0, 0],
       [1, 0, 0, 0, 0],
       [0, 0, 0, 1, 0]])

In [36]:
num_attribs = list(housing_num)

cat_attribs = ["ocean_proximity"]

num_pipeline = Pipeline([
        ('imputer', Imputer(strategy="median")),
        ('std_scaler', StandardScaler()),
    ])

cat_pipeline = Pipeline([
        ('selector', DataFrameSelector(cat_attribs)),
        ('lab_encoder',  MyLabelBinarizer()),
    ])

In [37]:
housing_num_tr = num_pipeline.fit_transform(housing_num)
housing_cat_tr = cat_pipeline.fit_transform(housing_cat)

In [38]:
housing_prepared = np.c_[housing_num_tr,housing_cat_tr]
housing_prepared

array([[-1.15604281,  0.77194962,  0.74333089, ...,  0.        ,
         0.        ,  0.        ],
       [-1.17602483,  0.6596948 , -1.1653172 , ...,  0.        ,
         0.        ,  0.        ],
       [ 1.18684903, -1.34218285,  0.18664186, ...,  0.        ,
         0.        ,  1.        ],
       ...,
       [ 1.58648943, -0.72478134, -1.56295222, ...,  0.        ,
         0.        ,  0.        ],
       [ 0.78221312, -0.85106801,  0.18664186, ...,  0.        ,
         0.        ,  0.        ],
       [-1.43579109,  0.99645926,  1.85670895, ...,  0.        ,
         1.        ,  0.        ]])

In [39]:
from sklearn.ensemble import RandomForestRegressor
forest_reg = RandomForestRegressor(random_state=42)
forest_reg.fit(housing_prepared, housing_labels)
housing_predictions = forest_reg.predict(housing_prepared)

In [40]:
from sklearn.model_selection import cross_val_score

forest_scores = cross_val_score(forest_reg, housing_prepared, housing_labels,
                                scoring="neg_mean_squared_error", cv=3)
forest_rmse_scores = np.sqrt(-forest_scores)
np.mean(forest_rmse_scores)

53522.73342861259